pfsimulator/colm/main/CoLM.F90

```diff
@@ -0,0 +1,958 @@
+         CALL hist_writeback_daemon ()
+      ELSE
+#endif
+
+      !IF (p_is_master) THEN
+      !   CALL system_clock (start_time)
+      !ENDIF
+
+      !casename     = DEF_CASE_NAME
+      !dir_landdata = DEF_dir_landdata
+      !dir_forcing  = DEF_dir_forcing
+      !dir_hist     = DEF_dir_history
+      !dir_restart  = DEF_dir_restart
+
+#ifdef SinglePoint
+      fsrfdata = trim(dir_landdata) // '/srfdata.nc'
+#ifndef URBAN_MODEL
+      CALL read_surface_data_single (fsrfdata, mksrfdata=.false.)
+#else
+      CALL read_urban_surface_data_single (fsrfdata, mksrfdata=.false., mkrun=.true.)
+#endif
+#endif
+
+      !deltim    = DEF_simulation_time%timestep
+      deltim    = dt*3600.d0
+      greenwich = DEF_simulation_time%greenwich
+      s_year    = DEF_simulation_time%start_year
+      s_month   = DEF_simulation_time%start_month
+      s_day     = DEF_simulation_time%start_day
+      s_seconds = DEF_simulation_time%start_sec
+      !e_year    = DEF_simulation_time%end_year
+      !e_month   = DEF_simulation_time%end_month
+      !e_day     = DEF_simulation_time%end_day
+      !e_seconds = DEF_simulation_time%end_sec
+      !p_year    = DEF_simulation_time%spinup_year
+      !p_month   = DEF_simulation_time%spinup_month
+      !p_day     = DEF_simulation_time%spinup_day
+      !p_seconds = DEF_simulation_time%spinup_sec
+
+      !spinup_repeat = DEF_simulation_time%spinup_repeat
+
+      CALL initimetype(greenwich) !set as true in namelist
+      CALL monthday2julian(s_year,s_month,s_day,s_julian)
+      !CALL monthday2julian(e_year,e_month,e_day,e_julian)
+      !CALL monthday2julian(p_year,p_month,p_day,p_julian)
+
+      sdate(1) = s_year; sdate(2) = s_julian; sdate(3) = s_seconds
+      !edate(1) = e_year; edate(2) = e_julian; edate(3) = e_seconds
+      !pdate(1) = p_year; pdate(2) = p_julian; pdate(3) = p_seconds
+
+      !@CY: build numpatch and planar_mask
+      allocate( counter(nx,ny) )
+      allocate( topo_mask(3,nx*ny) )            ! nx*ny >= numpatch
+      allocate( planar_mask(3,nx*ny) ) 
+      numpatch = 0
+      topo_mask = 0
+      topo_mask(3,:) = 1
+      planar_mask = 0
+      
+      do j = 1, ny
+         do i = 1, nx
+
+            ! i=tile(t)%col
+            ! j=tile(t)%row
+            counter(i,j) = 0
+            !clm(t)%topo_mask(3) = 1
+   
+            do k = nz, 1, -1 ! PF loop over z
+               l = 1+i + (nx+2)*(j) + (nx+2)*(ny+2)*(k)
+               if (topo(l) > 0) then
+                  counter(i,j) = counter(i,j) + 1
+                  if (counter(i,j) == 1) then
+                     numpatch = numpatch + 1
+                     topo_mask(1,numpatch) = k
+                     planar_mask(1,numpatch) = i
+                     planar_mask(2,numpatch) = j
+                     planar_mask(3,numpatch) = 1
+                  end if
+               endif
+   
+               if (topo(l) == 0 .and. topo(l+k_incr) > 0) topo_mask(3,numpatch) = k + 1
+   
+            enddo ! k
```

Comment:
This part of the code would reintroduce Issue #588.
The computation of the bottom index of the domain has been added to ParFlow, so you don't have to recompute it inside CoLM. Check how it was fixed for CLM:
https://github.com/parflow/parflow/blob/515956a2c9b81a05e680b823ffd571369d24d794/pfsimulator/clm/clm.F90#L336-L340
Instead of passing the topo to CoLM, you can just pass the top and bottom index of the domain.

Also, this reintroduction of the issue makes me wonder: how much more overlap is there between the two models CLM and CoLM? At a quick scan, I can't find any more glaring instances where this happens.
Is there a way to group common functionalities?